### PR TITLE
Driver: reorganize and rename

### PR DESCRIPTION
**Pull request by kpamnany**
# Description

We have a small mess of names in the driver and the configurations. `AtmosModel` too. In preparation for adding a diagnostics configuration step, this PR cleans this mess up a bit.

- `Common/ConfigTypes` centralizes definition of the various configuration types -- used both by the driver and by `AtmosModel`.
- Configurations are now created with `AtmosLESConfiguration` etc. (no underscores).

Also, break out solver configuration driver code into a new file.

See `examples/Atmos/dry_rayleigh_benard.jl` (or any of the experiment files) for the new names to be used.

# For review by CLIMA developers

- [X] There are no open pull requests for this already
- [X] CLIMA developers with relevant expertise have been assigned to review this submission
- [X] The code conforms to the [style guidelines](https://climate-machine.github.io/CLIMA/latest/CodingConventions.html) and has consistent naming conventions
- [X] This code does what it is technically intended to do (all numerics make sense physically and/or computationally)

### CODE REVIEW

**Comment by kpamnany** [comment]
@jkozdon: since you've been kindly helping with some cleanup of this code, I tagged you to take a look. :)

**Comment by kpamnany** [comment]
bors try

File: src/Driver/driver_configs.jl [hunk 1]
```diff
@@ -154,12 +156,12 @@ function Atmos_GCM_Configuration(
...
-    return DriverConfiguration(name, N, FT, array_type, solver_type, model,
-                               mpicomm, grid, numfluxnondiff, numfluxdiff,
-                               gradnumflux)
+    return DriverConfiguration(AtmosGCMConfigType(), name, N, FT, array_type,
+                               solver_type, model, mpicomm, grid,
+                               numfluxnondiff, numfluxdiff, gradnumflux)
 end
 
-function Ocean_BoxGCM_Configuration(
+function OceanBoxGCMConfiguration(
```
Comment (blallen) [inline src/Driver/driver_configs.jl]: `examples/OceanBoxGCM/homogenous_box.jl` and `ocean_gyre.jl` need to updated to match this, right?

File: src/Driver/solver_configs.jl [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+"""
+    CLIMA.setup_solver(t0, timeend, driver_config)
+
+Set up the DG model per the specified driver configuration and set up the ODE solver.
+"""
+function setup_solver(t0::FT, timeend::FT,
+                      driver_config::DriverConfiguration,
+                      init_args...;
+                      forcecpu=false,
+                      ode_solver_type=nothing,
```
Comment (jkozdon) [inline src/Driver/solver_configs.jl]:
```suggestion
                      ode_solver_type= driver_config.solver_type,
```

File: src/Driver/solver_configs.jl [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    # using the integral kernels -- these should have their own interface
+    update_aux!(dg, bl, Q, FT(0))
+    update_aux_diffusive!(dg, bl, Q, FT(0))
+
+    # if solver has been specified, use it
+    if ode_solver_type !== nothing
+        solver_type = ode_solver_type
+    else
+        solver_type = driver_config.solver_type
+    end
```
Comment (jkozdon) [inline src/Driver/solver_configs.jl]:
```suggestion
```

File: src/Driver/solver_configs.jl [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    dg = DGModel(bl, grid, numfluxnondiff, numfluxdiff, gradnumflux,
+                 modeldata=modeldata)
+    @info @sprintf("Initializing %s", driver_config.name)
+    Q = init_ode_state(dg, FT(0), init_args...; forcecpu=forcecpu)
+
+    # TODO: using `update_aux!()` to apply filters to the state variables and
+    # `update_aux_diffusive!()` to calculate the vertical component of velocity
+    # using the integral kernels -- these should have their own interface
+    update_aux!(dg, bl, Q, FT(0))
+    update_aux_diffusive!(dg, bl, Q, FT(0))
```
Comment (jkozdon) [inline src/Driver/solver_configs.jl]: Not sure this makes sense since diffusive hasn't been calculated yet.

File: src/Driver/solver_configs.jl [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+"""
+function DGmethods.courant(f, sc::SolverConfiguration; Q=sc.Q, dt = sc.dt)
+    DGmethods.courant(f, sc.dg, sc.dg.balancelaw, Q, dt)
+end
+
+"""
+    CLIMA.setup_solver(t0, timeend, driver_config)
+
+Set up the DG model per the specified driver configuration and set up the ODE solver.
+"""
```
Comment (jkozdon) [inline src/Driver/solver_configs.jl]: We may want to fill out the docs here for the arguments.

File: src/Driver/solver_configs.jl [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+    mpicomm::MPI.Comm
+    dg::DGModel
+    Q::MPIStateArray
+    t0::FT
+    timeend::FT
+    dt::FT
+    forcecpu::Bool
+    numberofsteps::Int
+    init_args
+    solver
```
Comment (jkozdon) [inline src/Driver/solver_configs.jl]: Do may want to comment on what these are?

File: src/Driver/solver_configs.jl (hunk 4 above)
Comment (blallen) [reply src/Driver/solver_configs.jl]: I think we can remove both of these now actually? I put them in for the HBModel, but looking in more detail, these aren't going to do anything here they won't do on the first time step.

File: src/Driver/driver_configs.jl (hunk 1 above)
Comment (kpamnany) [reply src/Driver/driver_configs.jl]: Oops.

**Comment by kpamnany** [comment]
> Nothing really major from my point of view. I've just been trying to clean up little rough edges as I run into them with testing. Haven't really given a ton of thought to the driver (just little mods on existing examples)

This is just a small cleanup. Just nice to get input from fresh eyes.

**Comment by kpamnany** [comment]
I also removed all trailing whitespaces so that git would stop complaining and being squelched. @simonbyrne: isn't there an automated way to do this?

File: src/Driver/solver_configs.jl (hunk 6 above)
Comment (kpamnany) [reply src/Driver/solver_configs.jl]: It's more or less a repeat of the parameters to `setup_solver()`. I don't know what the protocol is in such a case...

**Comment by kpamnany** [comment]
Any idea why this is happening @jkozdon? I don't see that I've changed anything that should affect this...?
```
Test Failed at D:\a\1\s\test\Driver\driver_test.jl:86
  Expression: isapprox(CFL_nondiff, CFL, rtol=0.03)
   Evaluated: isapprox(NaN, 0.4; rtol=0.03)
```

File: src/Driver/Driver.jl [hunk 7]
```diff
@@ -183,115 +190,8 @@ function init(; disable_gpu=false)
...
-                 diffusion_direction=diffdir)
-    @info @sprintf("Initializing %s", driver_config.name)
-    Q = init_ode_state(dg, FT(0), init_args...; init_on_cpu=init_on_cpu)
-
-    # TODO: using `update_aux!()` to apply filters to the state variables and
-    # `update_aux_diffusive!()` to calculate the vertical component of velocity
-    # using the integral kernels -- these should have their own interface
-    update_aux!(dg, bl, Q, FT(0))
-    update_aux_diffusive!(dg, bl, Q, FT(0))
-
```
Comment (jkozdon) [inline src/Driver/Driver.jl]: My guess is that this is the problem. You dropped the update_aux.

Comment (kpamnany) [reply src/Driver/Driver.jl]: Thank you!

**Comment by charleskawczynski** [comment]
@kpamnany Maybe don't modify the .dat files.

**Pull request merged**